### PR TITLE
[StructuralMechanicsApplication] Transition PR with MASS_FACTOR inclussion

### PR DESCRIPTION
**Pull request by loumalouomega**
In order to increase the time step during explicit simulations one technique is called mass scaling. This PR introduces the mass factor for a future PR with mass scaling computation.

### CODE REVIEW

**Review by msandre — DISMISSED** [review]

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.h [hunk 1]
```diff
@@ -64,9 +64,16 @@ double GetRayleighBeta(
     const ProcessInfo& rCurrentProcessInfo);
 
 /**
- * @brief Method to claculate the rayleigh damping-matrix
+ * @brief Method to returns the density to be consider for the mass-matrix computation
+ * @param rElement The Element for which the mass-matrix should be computed
+ * @return The density after apply the mass factor to the element
+ */
+double GetDensity(const Element& rElement);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.h]: Is this really only used in the computation of the MassMatrix?

=> the the method-name should be `GetDensityForMassMatrixComputation` or sth similar

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 2]
```diff
@@ -58,30 +63,51 @@ double GetRayleighAlpha(
...
+double GetDensity(const Element& rElement)
+{
+    // Getting the properties of the element
+    const auto& r_prop = rElement.GetProperties();
+
+    // Getting the density of the element
+    const double density = r_prop[DENSITY];
+
+    // Getting the mass factor
+    const double mass_factor = rElement.Has(MASS_FACTOR) ? rElement.GetValue(MASS_FACTOR) : r_prop.Has(MASS_FACTOR) ? r_prop.GetValue(MASS_FACTOR) : 1.0;
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp]: 1. can this be an elemental-data?

2. can you please follow the style of this file, which is more readable IMO?
~~~c


if (rElement.Has(MASS_FACTOR)) {
    return rElement.GetValue(MASS_FACTOR) * density;
} else if (r_prop.Has(MASS_FACTOR)) {
    return r_prop.GetValue(MASS_FACTOR) * density;
} else {
    return density;
}
~~~
3. Please add tests in the same way did such that the correct defaults are tested

**Review by philbucher — CHANGES_REQUESTED** [review]
It would be nice to also apply those changes in the shell-elements

Other than that I would like to briefly discuss this with @KlausBSautter (he is not in the office today, but tmr)

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.h (hunk 1 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.h]: Ok

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp]: Ok

**Comment by loumalouomega** [comment]
@philbucher done, but I prefer not  to update the shells, that use external utilities, and I don't feel confident to modify

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 3]
```diff
@@ -58,30 +63,55 @@ double GetRayleighAlpha(
...
+double GetDensityForMassMatrixComputation(const Element& rElement)
+{
+    // Getting the properties of the element
+    const auto& r_prop = rElement.GetProperties();
+
+    // Getting the density of the element
+    const double density = r_prop[DENSITY];
+
+    // Getting the mass factor
+    if (rElement.Has(MASS_FACTOR)) {
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp:103]: Can this be an elemental-data?

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 4]
```diff
@@ -58,30 +63,55 @@ double GetRayleighAlpha(
...
+        return rElement.GetValue(MASS_FACTOR) * density;
+    } else if (r_prop.Has(MASS_FACTOR)) {
+        return r_prop.GetValue(MASS_FACTOR) * density;
+    } else {
+        return density;
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp:108]: could the MASS_FACTOR also be globally defined in the ProcessInfo? => would this make sense?

**Review by philbucher — COMMENTED** [review]
I have two more questions, and please add a test in the same way I did

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp (hunk 3 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp:103]: It is supposed to be

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp (hunk 4 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp:108]: Can, but I prefer to be related with the properties (a material)

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp (hunk 3 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp:103]: ok

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp (hunk 4 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp:108]: ok

**Comment by philbucher** [comment]
> @philbucher done, but I prefer not to update the shells, that use external utilities, and I don't feel confident to modify

ok

**Comment by loumalouomega** [comment]
@philbucher is anything missing then?

**Comment by philbucher** [comment]
> @philbucher is anything missing then?

quoting myself: "and please add a test in the same way I did"

**Comment by loumalouomega** [comment]
Done @philbucher

**Comment by philbucher** [comment]
I mean a C-Test like [this](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/tests/cpp_tests/test_structural_mechanics_element_utilities.cpp#L27-L50)

**Comment by loumalouomega** [comment]
> I mean a C-Test like [this](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/tests/cpp_tests/test_structural_mechanics_element_utilities.cpp#L27-L50)

This tests as well, I don't feel the need to create a new test for something so simple

**Comment by philbucher** [comment]
I agree that it is simple but it ensures that the values are coming from the correct place (element has priority over property)
=> This is not covered by your test, so if someone decides to change it in the future we won't know bcs it is not tested

**Comment by loumalouomega** [comment]
> I agree that it is simple but it ensures that the values are coming from the correct place (element has priority over property)
> => This is not covered by your test, so if someone decides to change it in the future we won't know bcs it is not tested

The only think that changes in the test I added is the mass factor, so is properly tested, I think is properly tested. I don't want to add a cpp test that increases a lot the compilation time. Right now 1/4 compilations fails because of that, and adding things that increase the time is not very clever

**Comment by philbucher** [comment]
> > I agree that it is simple but it ensures that the values are coming from the correct place (element has priority over property)
> > => This is not covered by your test, so if someone decides to change it in the future we won't know bcs it is not tested
> 
> The only think that changes in the test I added is the mass factor, so is properly tested, I think is properly tested. I don't want to add a cpp test that increases a lot the compilation time. Right now 1/4 compilations fails because of that, and adding things that increase the time is not very clever

I don't think this is the best reason for not testing...
I can change the behavior by first checking the properties instead of first checking the element if a MASS_FACTOR is defined. This for example will not be detected by your test
=> Therefore I think it should be tested.
Even thought we have different opinions I will approve, since it is a detail and I will most likely not use this part of the code

**Review by philbucher — APPROVED** [review]

**Comment by loumalouomega** [comment]
> > > I agree that it is simple but it ensures that the values are coming from the correct place (element has priority over property)
> > > => This is not covered by your test, so if someone decides to change it in the future we won't know bcs it is not tested
> > 
> > 
> > The only think that changes in the test I added is the mass factor, so is properly tested, I think is properly tested. I don't want to add a cpp test that increases a lot the compilation time. Right now 1/4 compilations fails because of that, and adding things that increase the time is not very clever
> 
> I don't think this is the best reason for not testing...
> I can change the behavior by first checking the properties instead of first checking the element if a MASS_FACTOR is defined. This for example will not be detected by your test
> => Therefore I think it should be tested.
> Even thought we have different opinions I will approve, since it is a detail and I will most likely not use this part of the code

I think that my argument is valid, I don't want to increase more the compilation time, I already sacrificed my own application to be compiled by AppVeyor.

And if you look at the test I did I run the same code twice, there is only one variable that is changed between those runs, therefore it is the only reason the behaviour changed, then the code has been tested. Sorry if I don't deploy more cpp tests, but until @roigcarlo finishes the custom CI I don't want to add anything that can increase significantly the compilation time

**Comment by loumalouomega** [comment]
Thanks by the way

**Comment by philbucher** [comment]
Okok no worries:)

**Pull request merged**